File: dsa-05-main/cpp/day08/DoublyCircularList.cpp
#include <iostream>
using namespace std;

class DoublyCircularList;

class Node {
private:
	int data;
	Node *next;
	Node *prev;
public:
	Node(int val=0) {
		data = val;
		next = NULL;
		prev = NULL;
	}	
	friend class DoublyCircularList;
};

class DoublyCircularList {
private:
	Node *head;
	int count;
	
public:
	DoublyCircularList() {
		head = NULL;
		count = 0;
	}
	
	int getCount() {
		return count;
	}
	
	void display() {
		cout << "Fwd List: ";
		Node *trav = head;
		if(head != NULL) {
			do {
				cout << trav->data << ", ";
				trav = trav->next;
			}while(trav != head);
		}
		cout << endl;
		
		cout << "Rev List: ";
		if(head != NULL) {
			trav = head->prev;
			do {
				cout << trav->data << ", ";
				trav = trav->prev;
			}while(trav != head->prev);
		}
		cout << endl;
		
	}
	
	void addLast(int val) {
		// create new node<
		Node *newNode = new Node(val);
		// spl: if list is empty, add node at the start->
		if(head == NULL) {
			head = newNode;
			newNode->next = head;
			newNode->prev = head;
		} else {
			// mark last node as trav->
			Node *trav = head->prev;
			// add the node in between trav and head->
			newNode->next = head;
			newNode->prev = trav;
			trav->next = newNode;
			head->prev = newNode;
		}
		count++;
	}	
};

int main() {
	DoublyCircularList obj;
	obj.addLast(11);
	obj.addLast(22);
	obj.addLast(33);
	cout << "Count = " << obj.getCount() << endl;
	obj.display();
}

File: dsa-05-main/java/day08/doublycircularlist/DoublyCircularList.java
package sunbeam;

public class DoublyCircularList {

	static class Node {
		private int data;
		private Node next;
		private Node prev;
		public Node() {
			data = 0;
			next = null;
			prev = null;
		}
		public Node(int val) {
			data = val;
			next = null;
			prev = null;
		}
	}

	private Node head;
	
	public DoublyCircularList() {
		head = null;
	}
	
	public void displayFwd() {
		System.out.print("FWD LIST : ");
		if(head != null) {
			Node trav = head;
			do {
				System.out.print(trav.data + ", ");
				trav = trav.next;
			} while(trav != head);
		}
		System.out.println();
	}
	
	public void displayRev() {
		System.out.print("REV LIST : ");
		if(head != null) {
			Node trav = head.prev;
			do {
				System.out.print(trav.data + ", ");
				trav = trav.prev;
			} while(trav != head.prev);
		}
		System.out.println();
	}
	
	public void addLast(int val) {
		// create and init new node
		Node newNode = new Node(val);
		// if list is empty, make new node as first node and circular
		if(head == null) {
			head = newNode;
			newNode.next = head;
			newNode.prev = head;
		} else {
			// traverse till last node of the list
			Node trav = head.prev;
			// new node prev to last node and new node next to head
			newNode.prev = trav;
			newNode.next = head;
			// after last node add new node.
			trav.next = newNode;
			// head node prev to new node
			head.prev = newNode;
		}
	}
	
	public void addFirst(int val) {
		// create and init new node
		Node newNode = new Node(val);
		// if list is empty, make new node as first node and circular
		if(head == null) {
			head = newNode;
			newNode.next = head;
			newNode.prev = head;
		} else {
			// traverse till last node of the list
			Node trav = head.prev;
			// new node prev to last node and new node next to head
			newNode.prev = trav;
			newNode.next = head;
			// after last node add new node.
			trav.next = newNode;
			// head node prev to new node
			head.prev = newNode;
			// head to new node
			head = newNode;
		}
	}
	public void addAtPos(int val, int pos) {
		// homework
	}
	
	public void delFirst() {
		// homework
	}
	
	public void delAtPos(int pos) {
		// homework
	}
	
	public void delLast() {
		// homework
	}
	
	public void delAll() {
		head = null;
	}
}
















